posts/2023-12-10-2024-Edition-CFP.md

```diff
@@ -0,0 +1,112 @@
+so nobody ever has to worry about Edition-related incompatibility.
+
+In order to keep churn to a minimum, a new Edition of Rust is only released once every three years.
+We've had the [2015 Edition][2015], the [2018 Edition][2018], the [2021 Edition][2021],
+and soon, the 2024 Edition. And we could use your help!
+
+[2015]: https://doc.rust-lang.org/edition-guide/rust-2015/index.html
+
+[2018]: https://doc.rust-lang.org/edition-guide/rust-2018/index.html
+
+[2021]: https://doc.rust-lang.org/edition-guide/rust-2021/index.html
+
+## A call for proposals for the Rust 2024 Edition
+
+We know how much you love Rust, but let's be honest, no language is perfect,
+and Rust is no exception.
+So if you've got ideas for how Rust could be better if only that pesky stability guarantee
+weren't around, now's the time to share!
+Also note that potential Edition-related changes aren't just limited to the language itself:
+we'll also consider proposed changes to the standard library, Cargo, and rustfmt.
+
+Please keep in mind that the following criteria determine the sort of changes we're looking for:
+
+1. A change must be possible to implement without violating the strict properties
+listed in the prior section.
+Specifically, the ability of crates to have cross-Edition dependencies imposes restrictions
+on changes that would take effect across crate boundaries, e.g. the signatures of public APIs.
+However, we will occasionally discover that an Edition-related change
+[that was once thought to be impossible actually turns out to be feasible][change],
+so hope is not lost if you're not sure if your idea meets this standard;
+propose it just to be safe!
+
+[change]: https://doc.rust-lang.org/edition-guide/rust-2021/IntoIterator-for-arrays.html
+
+2. We strive to ensure that *nearly all* Edition-related changes can be applied
+to existing codebases automatically (via tools like `cargo fix`),
+in order to make upgrading to a new Edition as painless as possible.
+
+3. Even if an Edition *could* make any given change, [that doesn't mean that it should][jeff].
+We're not looking for hugely-invasive changes or things that would fundamentally
+alter the character of the language.
+Please focus your proposals on things like fixing obvious bugs, changing annoying behavior,
+unblocking future feature development, and making the language easier and more consistent.
+
+[jeff]: https://www.youtube.com/watch?v=9nazm3_OXac
+
+As an example, in the 2015 Edition a dynamically-dispatched trait object
+has a type that looks like `Box<Foo>`.
+In the 2018 Edition, in order to distinguish this from the then-new,
+statically-dispatched `impl Trait` syntax, trait objects began to require the `dyn` keyword,
+so `Box<Foo>` became `Box<dyn Foo>`.
+To make this migration automatic, `cargo fix` gained the ability to rewrite the types
+of trait objects,
+as well as to rewrite identifiers in case any code out there was using `dyn` as a variable name.
```

Comment:
If you want to have a different example, here's a suggestion:

```suggestion
As an example, the `async` and `await` keywords were added in the 2018 Edition in preparation for adding async support.
In the 2015 Edition, it was possible to have identifiers using those keywords like `let async = 1;`.
Adding the keywords would have been a breaking change without the Edition system.
With Edition support, the new keywords only exist in the 2018 and newer Editions.
Projects still declared as using the 2015 Edition continue to work.
To automatically migrate to the 2018 edition, `cargo fix --edition` will translate the code to use the raw-identifier syntax such as `let r#async = 1;` which allows it to continue to use the same identifier without conflicting with the new keyword. 
```

Reply:
I'm fine with the existing example, this is just an alternate per mara's suggestion above.

I don't think there is an existing example that really exemplifies a substantial syntax change. All of the syntax changes are relatively subtle.

`dyn Trait` was implemented as a weak keyword, which was backwards-compatible and not related to an edition. The edition only *rejects* the old syntax under the intent that we wanted to force a stylistic change.

Anonymous trait function parameters is also like `dyn Trait`. It only rejects old syntax to force a style change. Same with `start...end` patterns.

The only syntax changes I can think of that aren't rejection of existing syntax for style reasons is async and literal prefixes which will be relevant in 1.76 in February with c-string literals, which would otherwise be a breaking change without edition support.

Reply:
C-string literals sound like they could make for a good example, since that also serves as a preview of Rust 2024. :P

Reply:
Just FYI, c-string literals are in 2021 (because that's where it was reserved), not 2024.